### PR TITLE
Dependencies catch up

### PR DESCRIPTION
**Pull request by dmdashenkov**
In this PR we:
 - update Spine dependencies to the latest published versions;
 - update Datastore API dependency to `v1.2.1` (the freshest);
 - perform misc clean-up in the build scripts; migrate to a separate `ext.gradle` for version declaration.

The framework version is set to `0.9.37-SNAPSHOT`.

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -165,3 +165,43 @@ startDatastore.doLast {
...
+                                      changing: true,
+                                      excludes)
+    } else {
+        dependencyHandler.testCompile(group: 'io.spine',
+                                      name: artifactId,
+                                      version: version,
+                                      changing: true,
+                                      excludes)
+    }
+
```
Comment (armiol) [inline build.gradle]: Please remove this line.

File: ext.gradle [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+/**
+ *  The versions of the libraries used.
+ *
+ *  This file is used wherever the library versions are required.
+ */
+
+final SPINE_VERSION = "0.9.36-SNAPSHOT"
+
+ext {
+    VERSION = SPINE_VERSION
```
Comment (armiol) [inline ext.gradle]: Please have the same camel-case format as in `core-java`.

File: ext.gradle [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    BASE_VERSION = "0.9.35-SNAPSHOT"
+    PROTOBUF_VERSION = '3.3.0'
+    SLf4J_VERSION = "1.7.21"
+    PROTOBUF_DEPENDENCY = "com.google.protobuf:protoc:${project.PROTOBUF_VERSION}"
+    MAVEN_REPOSITORY_URL = 'http://maven.teamdev.com/repository/spine'
+    MAVEN_SNAPSHOT_REPOSITORY_URL = 'http://maven.teamdev.com/repository/spine-snapshots'
+
+    GUAVA_VERSION = "20.0"
+
+    IS_WINDOWS = org.gradle.internal.os.OperatingSystem.current().isWindows()
```
Comment (armiol) [inline ext.gradle]: The idea of `ext.gradle` is to externalize the reusable dependency list, which is being changed on a regular basis. This variable does not belong to `ext.gradle`.

File: build.gradle [hunk 4]
```diff
@@ -165,3 +165,43 @@ startDatastore.doLast {
...
+/**
+ * Adds a compile-time dependency on a Spine module.
+ *
+ * @param dependencyHandler the project dependency handler (basically {@code project.dependencies})
+ * @param artifactId the ID of the Spine module artifact
+ * @param version the Spine module version
+ * @param testScope tag parameter: {@code true} if the dependency should be added only for the tests
+ *                  (e.g. test utils modules`), {@code false} otherwise
+ */
+void dependOnSpine(final dependencyHandler,
```
Comment (armiol) [inline build.gradle]: Let's not have that at all. It's way more lines than having just a list of dependencies explicitly.

Also, if you want to exclude some of transitive dependencies, you could have a collection in the same way and pass it to `excludes: ` section in each case. Or use global `configurations.all { ...}`.

File: build.gradle [hunk 5]
```diff
@@ -1,24 +1,26 @@
...
     publishPlugin = "$rootDir/script/publish.gradle"
     credentialsPropertyFile = 'credentials.properties'
     projectsToPublish = ['datastore']
+
+    /**
+     * A map of the Spine project IDs to their versions.
+     */
+    spineDependencies = ['spine-server': VERSION,
+                         'spine-client': VERSION,
+                         'spine-core': VERSION,
```
Comment (armiol) [inline build.gradle]: I think the `spine-core` and `spine-base` dependencies should not be here, as you'll get those anyway.

File: datastore/src/main/java/io/spine/server/storage/datastore/DatastoreStorageFactory.java [hunk 6]
```diff
@@ -151,12 +152,12 @@ public StandStorage createStandStorage() {
      * {@inheritDoc}
      */
     @Override
-    public <I> ProjectionStorage<I> createProjectionStorage(Class<? extends Entity<I, ?>> aClass) {
-        final DsRecordStorage<I> recordStorage = (DsRecordStorage<I>) createRecordStorage(aClass);
+    public <I> ProjectionStorage<I> createProjectionStorage(Class<? extends Projection<I, ?, ?>> projectionClass) {
+        final DsRecordStorage<I> recordStorage = (DsRecordStorage<I>) createRecordStorage((Class<? extends Entity<I, ?>>) projectionClass);
```
Comment (armiol) [inline datastore/src/main/java/io/spine/server/storage/datastore/DatastoreStorageFactory.java]: Please fit 100 symbols wide.

**Comment by dmdashenkov** [comment]
@armiol, PTAL.

The build still fails since the required core version is not in the artifactory yet.

File: datastore/src/test/java/io/spine/server/storage/datastore/DsRecordStorageShould.java [hunk 7]
```diff
@@ -250,9 +259,11 @@ public void write_and_read_records_with_lifecycle_flags() {
         assertEquals(record, restoredRecord);
     }
 
+    // ----- Test Entity Types ------
```
Comment (armiol) [inline datastore/src/test/java/io/spine/server/storage/datastore/DsRecordStorageShould.java]: Please the section title according to our guidelines. See the `core-java` wiki.

File: gradle/wrapper/gradle-wrapper.properties [hunk 8]
```diff
@@ -3,4 +3,4 @@ distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-3.5-bin.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-4.0-bin.zip
```
Comment (armiol) [inline gradle/wrapper/gradle-wrapper.properties:6]: I think JAR file should be updated as well. Maybe, the `.bat`/`.sh` scripts also.

**Review by armiol — CHANGES_REQUESTED** [review]
@dmdashenkov please see my comments.

File: gradle/wrapper/gradle-wrapper.properties (hunk 8 above)
Comment (dmdashenkov) [reply gradle/wrapper/gradle-wrapper.properties:6]: Done.

**Comment by dmdashenkov** [comment]
@armiol, PTAL.

**Comment by dmdashenkov** [comment]
@armiol, PTAL.

File: datastore/src/main/java/io/spine/server/storage/datastore/DatastoreStorageFactory.java [hunk 9]
```diff
@@ -231,9 +235,11 @@ public static Builder newBuilder() {
     public static class Builder {
 
         private static final String DEFAULT_NAMESPACE_ERROR_MESSAGE =
-                "Datastore namespace should not be configured explicitly for a multitenant storage.";
+                "Datastore namespace should not be configured explicitly for a multitenant " +
```
Comment (armiol) [inline datastore/src/main/java/io/spine/server/storage/datastore/DatastoreStorageFactory.java]:
```
"Datastore namespace should not be configured explicitly" +
"for a multitenant storage"
```

File: datastore/src/main/java/io/spine/server/storage/datastore/DatastoreStorageFactory.java [hunk 10]
```diff
@@ -231,9 +235,11 @@ public static Builder newBuilder() {
...
+                        "storage.";
         private static final String REDUNDANT_TENANT_ID_CONVERTER_ERROR_MESSAGE =
-                "Setting a custom NamespaceToTenantIdConverter to a single-tenant storage factory is redundant.";
+                "Setting a custom NamespaceToTenantIdConverter to a single-tenant storage factory" +
```
Comment (armiol) [inline datastore/src/main/java/io/spine/server/storage/datastore/DatastoreStorageFactory.java]:
```
"Setting a custom NamespaceToTenantIdConverter is not allowed" + 
" for a single-tenant storage factory."
```

File: datastore/src/test/java/io/spine/server/datastore/TestEnvironment.java [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.spine.server.datastore;
+
+import static java.lang.System.getenv;
+
+/**
+ * A utility class for investigating the test environment at runtime.
```
Comment (armiol) [inline datastore/src/test/java/io/spine/server/datastore/TestEnvironment.java]: `for analyzing`/`for detecting`

File: datastore/src/test/java/io/spine/server/storage/datastore/given/Given.java [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+
+/**
+ * A container for the Spine test environment constants.
+ *
+ * @author Dmytro Dashenkov
+ */
+@Internal
+public final class Given {
+
+    public static final String TEST_PROJECT_ID_VALUE = "spine-dev";
```
Comment (armiol) [inline datastore/src/test/java/io/spine/server/storage/datastore/given/Given.java]: Please expose those as static methods instead.

**Review by armiol — APPROVED** [review]
@dmdashenkov LGTM except for a few things. Please address those and feel free to merge.

**Comment by armiol** [comment]
@dmdashenkov LGTM. I approved this PR yesterday :)

**Pull request merged**